packages/frontend/src/system-integration/notifications.ts

```diff
@@ -260,41 +308,85 @@ async function flushNotifications(accountId: number) {
   )
     .filter(e => e.muted)
     .map(e => e.id)
-  if (mutedChats.length > 0) {
-    // some chats are muted
-    log.debug(`ignoring notifications of ${mutedChats.length} muted chats`)
-  }
-  notifications = notifications.filter(notification => {
-    if (mutedChats.includes(notification.chatId)) {
-      // muted chat
-      log.debug('notification ignored: chat muted', notification)
-      return false
-    } else {
-      return true
-    }
-  })
 
-  if (notifications.length > notificationLimit) {
+  const filteredNotifications = (
+    await Promise.all(
+      notifications.map(async notification => {
+        if (mutedChats.includes(notification.chatId)) {
+          // muted chat - only show if it's a mention and mentions are enabled
+          const isMention = await notificationIsMention(accountId, notification)
+          return isMention ? notification : null
+        } else {
+          log.debug('notification on not muted chat:', notification)
+          return notification
+        }
+      })
+    )
+  ).filter(notification => notification !== null)
+
+  if (filteredNotifications.length > notificationLimit) {
     showGroupedNotification(accountId, notifications)
   } else {
     for (const {
       chatId,
       messageId,
-      isWebxdcInfo,
+      notificationType,
       eventText,
-    } of notifications) {
+      contactId,
+    } of filteredNotifications) {
       await showNotification(
         accountId,
         chatId,
         messageId,
-        isWebxdcInfo,
-        eventText
+        notificationType,
+        eventText,
+        contactId
       )
     }
   }
   notificationLimit = NORMAL_LIMIT
 }
 
+/**
+ * if isMentionsEnabled returns true
+ * if the notification is a mention
+ */
+async function notificationIsMention(
+  accountId: number,
+  notification: QueuedNotification
+) {
+  if (!SettingsStoreInstance.state?.desktopSettings.isMentionsEnabled) {
+    log.info('allowIfMention: mentions are disabled')
+    return false
+  } else if (notification.notificationType === NOTIFICATION_TYPE.WEBXDC_INFO) {
+    log.info('mention detected: webxdc-info notification')
+    return true
+  } else {
+    const message = await BackendRemote.rpc.getMessage(
+      accountId,
+      notification.messageId
+    )
+    if (notification.notificationType === NOTIFICATION_TYPE.REACTION) {
+      if (message.sender.id === C.DC_CONTACT_ID_SELF) {
+        log.info('mention detected: reaction to own message')
+        return true
+      }
+    }
+    if (message.quote && message.quote.kind === 'WithMessage') {
+      const quote = await BackendRemote.rpc.getMessage(
+        accountId,
+        message.quote.messageId
+      )
+      if (quote.sender.id === C.DC_CONTACT_ID_SELF) {
+        log.info('mention detected: answer to own message')
+        return true
+      }
+    }
+    log.debug('ignoring notification on muted chat')
+    return false
+  }
```

Comment:
```suggestion
  if (!SettingsStoreInstance.state?.desktopSettings.isMentionsEnabled) {
    log.info('allowIfMention: mentions are disabled')
    return false
  }

  if (notification.notificationType === NOTIFICATION_TYPE.WEBXDC_INFO) {
    log.info('mention detected: webxdc-info notification')
    return true
  }

  const message = await BackendRemote.rpc.getMessage(
    accountId,
    notification.messageId
  )

  if (
    notification.notificationType === NOTIFICATION_TYPE.REACTION &&
    message.sender.id === C.DC_CONTACT_ID_SELF
  ) {
    log.info('mention detected: reaction to own message')
    return true
  }

  if (
    message.quote &&
    message.quote.kind === 'WithMessage' &&
    (await BackendRemote.rpc.getMessage(accountId, message.quote.messageId))
      .sender.id === C.DC_CONTACT_ID_SELF
  ) {
    log.info('mention detected: answer to own message')
    return true
  }

  log.debug('ignoring notification on muted chat')
  return false
```

Less nesting. See if you like it more.

Reply:
I agree the nesting be reduced a bit